src/rpc/daemon_handler.cpp

```diff
@@ -0,0 +1,881 @@
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "daemon_handler.h"
+
+// likely included by daemon_handler.h's includes,
+// but including here for clarity
+#include "cryptonote_core/cryptonote_core.h"
+#include "cryptonote_basic/cryptonote_format_utils.h"
+#include "cryptonote_protocol/blobdatatype.h"
+#include "ringct/rctSigs.h"
+
+namespace cryptonote
+{
+
+namespace rpc
+{
+
+  void DaemonHandler::handle(const GetHeight::Request& req, GetHeight::Response& res)
+  {
+    res.height = m_core.get_current_blockchain_height();
+
+    res.status = Message::STATUS_OK;
+  }
+
+  void DaemonHandler::handle(const GetBlocksFast::Request& req, GetBlocksFast::Response& res)
+  {
+    std::list<std::pair<blobdata, std::list<blobdata> > > blocks;
+
+    if(!m_core.find_blockchain_supplement(req.start_height, req.block_ids, blocks, res.current_height, res.start_height, COMMAND_RPC_GET_BLOCKS_FAST_MAX_COUNT))
+    {
+      res.status = Message::STATUS_FAILED;
+      res.error_details = "core::find_blockchain_supplement() returned false";
+      return;
+    }
+
+    res.blocks.resize(blocks.size());
+    res.output_indices.resize(blocks.size());
+
+    //TODO: really need to switch uses of std::list to std::vector unless
+    //      it's a huge performance concern
+
+    auto it = blocks.begin();
+
+    uint64_t block_count = 0;
+    while (it != blocks.end())
+    {
+      cryptonote::rpc::block_with_transactions& bwt = res.blocks[block_count];
+
+      if (!parse_and_validate_block_from_blob(it->first, bwt.block))
+      {
+        res.blocks.clear();
+        res.output_indices.clear();
+        res.status = Message::STATUS_FAILED;
+        res.error_details = "failed retrieving a requested block";
+        return;
+      }
+
+      std::list<transaction> txs;
+      for (const auto& blob : it->second)
+      {
+        txs.resize(txs.size() + 1);
+        if (!parse_and_validate_tx_from_blob(blob, txs.back()))
+        {
+          res.blocks.clear();
+          res.output_indices.clear();
+          res.status = Message::STATUS_FAILED;
+          res.error_details = "failed retrieving a requested block";
```

Comment:
fair enough; I put block because it's in the get blocks rpc, but tx makes sense, context should be known.